File: src/main/java/com/dsa/problems/scaler/sort/count_sort.java
package com.dsa.problems.scaler.sort;

import java.util.ArrayList;
import java.util.Arrays;

public class count_sort {
  /**
   * Q. Count Sort
   *
   * Given an array A. Sort this array using Count Sort Algorithm and return the sorted array.
   *
   * Problem Constraints
   * 1 <= |A| <= 10^5
   * 1 <= A[i] <= 10^5
   */
  private static void countSort(ArrayList<Integer> A) {
    final int freqSize = (int)(Math.pow(10, 5) + 1);
    ArrayList<Integer> freq = new ArrayList<>();
    for(int i = 0; i < freqSize; i++) {
      freq.add(i, 0);
    }

    for(int i = 0; i < A.size(); i++) {
      int temp = A.get(i);
      freq.set(temp, freq.get(temp) + 1);
    }

    int j = 0;
    for(int i = 0; i < freqSize; i++) {
      for(int z = 0; z < freq.get(i); z++) {
        A.set(j++, i);
      }
    }
  }

  public static void main(String[] args) {
    ArrayList<Integer> A = new ArrayList<>(Arrays.asList(2,3,5,6,4,2,5,2,4,3,2,5,3,9,7,2,5));
    countSort(A);
    System.out.println(A);
  }
}

File: problems/scaler/sort/count_sort.js
/**
 * Q. Count Sort
 *
 * Given an array A. Sort this array using Count Sort Algorithm and return the sorted array.
 *
 * Problem Constraints
 * 1 <= |A| <= 105
 * 1 <= A[i] <= 105
 */
function soln(A) {
  let freq = new Array(100001).fill(0);
  for (const a of A) {
    freq[a]++;
  }

  let i = 0;
  for (let j = 0; j < 100001; j++) {
    for (let x = 0; x < freq[j]; x++) {
      A[i] = j;
      i++;
    }
  }

  return A;
}

console.log(soln([1, 3, 1]));
